### PR TITLE
Proposition: After the OpenCL branch maturing for over 3 years, maybe it's time to give Caffe a bump

### PR DESCRIPTION
**Pull request by naibaf7**
The OpenCL branch has been maturing for 3 years. In this time, a lot of features have been accumulated that are absent in the main branch. Notably, there are many features (and many more to come) that would differentiate Caffe from other frameworks such as TensorFlow, Caffe2 and Torch in a positive way. Caffe has been lacking such features lately.

I'm also standing on a crossroad now, where adding more features to my branch would ultimately make it too difficult to merge upstream changes from the master branch into the opencl branch.

A selection of core features:
- Virtual pointers and complete device abstraction.
- Multiplexer supporting different BLAS libraries on GPUs and CPUs.
- A single, device-abstracted code path for the GPU layers.
- Runtime compiled kernels for CUDA and OpenCL with a SQLite3 kernel cache.
- LibDNN with support for convolution, pooling and a complete BLAS, that can act as a drop-in replacement for cuDNN and runs on all OpenCL and CUDA GPUs.
- Extended Python interface.
- Vastly improved NetSpec.
- Quantized data types (INT8, INT16) and half-precision float (FP16).
- Quantization with support of pseudo-quantization, quantization parameter estimation and mixed-precision neural networks.
- Gemmlowp support on the CPU (INT8 inference).
- Topological MALIS-loss layer.
- Mixture-of-experts (MOE) layer.
- Explicit support for AMD, nVidia, ARM Mali and Intel GPUs.
- CPU OpenCL support.
- Extended OpenMP support.
- Dropping the Makefile build system.
- CMake cross-compile support for Android and ARM (Raspberry Pi, Asus Tinkerboard)
- Reduced memory inference.
- Runtime compiled kernels means no more conflicts between NVCC and host compiler, making builds much faster and easier.
- Forward and backward compatibility to existing networks and trained weights.

Upcoming features:
- Raspberry Pi VideoCore IV support
- Writing GPU layers in Python

Currently broken features:
- NCCL
- Multi-GPU
- HDF5 (in some cases)
- Some runtests (easy to fix)

Missing/incomplete features:
- Some unit tests for quantized types are not implemented yet
- Quantization only implemented for some layers (LeNet and AlexNet work quantized).

So I hope we can get some maintainers to help me hunt down the remaining few bugs in this branch and make it ready for a merge with the master branch (maybe until December?).

I hope this finds consideration among other Caffe maintainers. I will follow up with all changes and additions made to this branch once core maintainers will agree to go down this path for the future of Caffe.

### CODE REVIEW

**Comment by skn123** [comment]
Fabian, unable to build this branch

**Comment by shelhamer** [comment]
@naibaf7 Thank you for dedicating so much effort and engineering to OpenCL Caffe! This branch has only improved since the old times of #2610. Throughout the discussion and comments you have always contributed code, and that takes a lot of time and thought.

I have clearly not had the bandwidth for Caffe development that I once did. I still have my sights on a 1.1 to include module layers (and by extension Halide layers), because these are excellent directions that I was always interested in—among many other good directions, of course—but it has been a long while now. While I still hope to rally the time to take care of last bits of grooming and merge those, I do not expect to be able to contribute as necessary to make a difference in the ~100k line patch that you are hammering on here.

That said, do not let that be an obstacle to your hacking! The only constraint I would advise is that existing capabilities not go missing, like multi-GPU training, HDF5 IO, and so forth. Whether this line of development is a branch, fork, or merged in master can be up to you and the others who keep pushing, pulling, reviewing, and of course brewing :coffee:

**Comment by TechnikEmpire** [comment]
"Thanks for all the hard work" - Lets it rot as a dangling PR forever.

**Comment by Coderx7** [comment]
@naibaf7 : what happened? whats the ultimate decision ?

**Comment by naibaf7** [comment]
@TechnikEmpire 
Currently I do not have time to continue developing it.
Neither seems interest big enough to do so, and since this was an academic exercise, I doubt it will go on like it is.
However, I am looking into porting the Caffe OpenCL code into LibDNN, and then trying to get PyTorch onto that. But it's not sure yet.
Don't expect any news short-time.

**Comment by TechnikEmpire** [comment]
Thanks for your contribution @naibf7. It's unfortunate but Nvidia has a strangle hold on deep learning, and now Intel has joined the ranks, even destroying opencv and making it depend on propertary acceleration that doesn't even work on amd cpus.

People like you with your contributions are opening these technologies to the general public while others hold back progress to line their corporate pockets. Will watch your personal repos. Thanks again.

**Comment by skn123** [comment]
@naibaf7 This was indeed a welcome commit. Can you please check the PR That I made? If a few issues related to "double" is solved then the code will be fully functional. However, there is a HSA version of caffe being developed at ROCm! I am waiting for that. The older version of the code is also there in my personal repo and it should work without any hitch.

**Comment by dragonQian** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
